File: 138__Copy_List_with_Random_Pointer.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution(object):

    def __init__(self):
        self.visited = {}

    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """

        if head is None:
            return None

        if head in self.visited:
            return self.visited[head]

        node = RandomListNode(head.label)
        self.visited[head] = node

        node.next = self.copyRandomList(head.next)
        node.random = self.copyRandomList(head.random)

        return node

File: Java/138__Copy_List_with_Random_Pointer.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node next;
    public Node random;

    public Node() {}

    public Node(int _val,Node _next,Node _random) {
        val = _val;
        next = _next;
        random = _random;
    }
};
*/
class Solution {
    HashMap<Node, Node> node_map = new HashMap<>();
    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }
        if (node_map.containsKey(head)) {
            return node_map.get(head);
        }
        Node result = new Node(head.val, null, null);
        node_map.put(head, result);
        result.next = this.copyRandomList(head.next);
        result.random = this.copyRandomList(head.random);
        
        return result;
    }
}
